### PR TITLE
feat (customer_metadata): handle metadata for graphQL create and update

### PR DESCRIPTION
**Pull request by lovrocolic**
## Context

Currently we are not able to persist additional customer data

## Description

This PR handles metadata from graphQL for create and update actions

### CODE REVIEW

File: app/graphql/types/customer_metadata/input.rb [hunk 1]
```diff
@@ -0,0 +1,15 @@
+# frozen_string_literal: true
+
+module Types
+  module CustomerMetadata
```
Comment (rsempe) [inline app/graphql/types/customer_metadata/input.rb:4]: What about `Types::Customers::Metadata::Input` instead?

File: app/services/customers/create_service.rb [hunk 2]
```diff
@@ -86,6 +91,14 @@ def create(**args)
 
     private
 
+    def create_metadata(customer, args)
```
Comment (rsempe) [inline app/services/customers/create_service.rb:94]:
```suggestion
    def create_metadata(customer:, args:)
```

File: app/graphql/types/customer_metadata/input.rb [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+    class Input < Types::BaseInputObject
+      graphql_name 'CustomerMetadataInput'
+
+      argument :id, ID, required: false
+      argument :key, String, required: true
+      argument :value, String, required: true
+
+      argument :display_in_invoice, Boolean, required: true
```
Comment (rsempe) [inline app/graphql/types/customer_metadata/input.rb:12]: Is `display_on_invoice` wording not better?

File: app/graphql/types/customer_metadata/input.rb (hunk 1 above)
Comment (lovrocolic) [reply app/graphql/types/customer_metadata/input.rb]: changed 👍

File: app/services/customers/create_service.rb (hunk 2 above)
Comment (lovrocolic) [reply app/services/customers/create_service.rb]: updated

File: app/services/customers/update_service.rb [hunk 4]
```diff
@@ -55,7 +55,12 @@ def update(**args)
 
         # NOTE: external_id is not editable if customer is attached to subscriptions
         customer.external_id = args[:external_id] if customer.editable? && args.key?(:external_id)
-        customer.save!
+
+        ActiveRecord::Base.transaction do
+          customer.save!
+
+          process_metadata(customer, args[:metadata]) unless args[:metadata].nil?
```
[the remark this answers is not in the document]
Comment (lovrocolic) [reply app/services/customers/update_service.rb]: extracted 👍

File: app/graphql/types/customer_metadata/input.rb (hunk 3 above)
Comment (lovrocolic) [reply app/graphql/types/customer_metadata/input.rb]: actually I have checked those two versions; both are grammatically correct but from what I have seen syntax `in invoice` is used more 😅

File: app/services/customers/update_service.rb (hunk 4 above)
[the remark this answers is not in the document]
Comment (lovrocolic) [reply app/services/customers/update_service.rb]: @rsempe Extracted, but this service call is a bit changed in another PR (based on our conversation there)

File: app/services/customers/metadata/update_service.rb [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+
+            next
+          end
+
+          created_metadata = create_metadata(payload_metadata)
+          created_metadata_ids.push(created_metadata.id)
+        end
+
+        # NOTE: Delete metadata that are no more linked to the customer
+        sanitize_charges(hash_metadata, created_metadata_ids)
```
Comment (vincent-pochet) [inline app/services/customers/metadata/update_service.rb:29]: Shouldn't we name this one `sanitize_metadata`?

**Review by vincent-pochet — APPROVED** [review]

**Review by rsempe — APPROVED** [review]

File: app/services/customers/metadata/update_service.rb (hunk 5 above)
Comment (lovrocolic) [reply app/services/customers/metadata/update_service.rb]: Oh, thanks for catching!

**Pull request merged**